### PR TITLE
Fix broken undo history stack for Pattern Overrides

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

The undo history is a bit broken for Pattern Overrides. This PR tries to fix it.

## Why?

For better editing experience.

## How?

This PR removes the `syncDerivedBlockAttributes` action introduced in the original PR (#56235) an instead introduce a new private higher-order action `syncDerivedUpdates`.

The signature of `syncDerivedUpdates` looks like:
```ts
syncDerivedUpdates( callback: () => void ): void;
```

Any updates inside the callback should not push additional undo history stacks nor should be persistent. This serves as a escape hatch to make updates to the block editor completely invisible to the core's undo manager. This means we can call `setAttributes` to set the `overrides` of the pattern block without worrying that it might break the history stack.

Example usage:
```js
const { syncDerivedUpdates } = unlock( registry.dispatch( blockEditorStore ) );

syncDerivedUpdates( () => {
  setAttributes({ overrides });
} );
```

The goal is to make something like this stable enough so that third-party developers can leverage it too. We probably can use it solve some of the longstanding undo bugs too (https://github.com/WordPress/gutenberg/issues/8119#issuecomment-1185238067, https://github.com/WordPress/gutenberg/issues/41031#issuecomment-1188544191).

> [!NOTE]
> This API currently doesn't work 100% of the time for other blocks so it should remain private for now. The top priority of this PR is to fix the existing issue for Pattern Overrides. We can work on finalizing the API across different use cases in the future.

## Testing Instructions

1. Enable the "Partial syncing patterns gutenberg experiment.
2. Go to Site Editor -> Patterns -> Create a synced pattern.
3. Add a paragraph and go to the block inspector sidebar -> Advanced -> Check the "Synced attributes" checkbox.
4. Save the pattern.
5. Go to edit a post.
6. Insert the pattern.
7. Make some changes to the enabled paragraph and test the undo history.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/7753001/0f95849b-0076-48ec-9687-98fb5bc46a1d

### CODE REVIEW

File: packages/block-editor/src/store/private-actions.js [hunk 1]
```diff
@@ -291,10 +296,30 @@ export function deleteStyleOverride( id ) {
...
+			} );
+			callback();
+			dispatch( {
+				type: 'SET_EXPLICIT_PERSISTENT',
+				isPersistentChange: undefined,
+			} );
+
+			// Ignore pushing undo stack for the updated blocks.
+			const updatedBlocks = select.getBlocks();
+			undoIgnoreBlocks.add( updatedBlocks );
```
Comment (Mamaduka) [inline packages/block-editor/src/store/private-actions.js:322]: It would be nice to be more explicit here and store only updated blocks instead of every block. Probably something for the future and before API is stable.

File: packages/editor/src/components/document-outline/test/index.js [hunk 2]
```diff
@@ -17,10 +17,6 @@ import {
  */
 import { DocumentOutline } from '../';
 
-jest.mock( '@wordpress/block-editor', () => ( {
-	BlockTitle: () => 'Block Title',
-} ) );
-
```
Comment (Mamaduka) [inline packages/editor/src/components/document-outline/test/index.js]: Is this related to the changes here?

**Review by Mamaduka — COMMENTED** [review]
Briefly tested and read through the code. I think this is an interesting idea.

Will try to do more thorough testing later this week.

P.S. Tried to use the `state.blocks.isPersistentChange` flag when working on a fix for the ToC block, but I couldn't get it working 😅

File: packages/block-editor/src/store/private-actions.js (hunk 1 above)
Comment (kevin940726) [reply packages/block-editor/src/store/private-actions.js:322]: It's storing the updated blocks reference in a `WeakSet` so it shouldn't create extra memory nor leak it. In a way it's actually faster to not do any transformation here 😆. We can probably achieve the same thing with a flag in the store, like what `isPersistentChange` does, but the higher-order reducer makes it unnecessarily hard 😅. We can revisit this part in a future API design.

File: packages/editor/src/components/document-outline/test/index.js (hunk 2 above)
Comment (kevin940726) [reply packages/editor/src/components/document-outline/test/index.js]: Not anymore! 😛

**Comment by kevin940726** [comment]
Thanks for testing it out! ❤️

> Tried to use the `state.blocks.isPersistentChange` flag when working on a fix for the ToC block, but I couldn't get it working 😅

I tried using `syncDerivedUpdates` in the ToC block but seems like the `subscribe` function messes things up a bit so it currently doesn't work (hence the note in the PR description). Namely, the timing of calling `registry.subscribe` globally (without the second argument) behaves differently from only subscribing to a single store. We can look into that in the future. 🤔

File: packages/block-editor/src/store/private-actions.js [hunk 3]
```diff
@@ -291,10 +296,30 @@ export function deleteStyleOverride( id ) {
 	};
 }
 
-export function syncDerivedBlockAttributes( clientId, attributes ) {
-	return {
-		type: 'SYNC_DERIVED_BLOCK_ATTRIBUTES',
-		clientIds: [ clientId ],
-		attributes,
+/**
+ * A higher-order action that mark every change inside a callback as "non-persistent"
```
Comment (glendaviesnz) [inline packages/block-editor/src/store/private-actions.js:300]:
```suggestion
 * A higher-order action that marks every change inside a callback as "non-persistent"
 * and ignores pushing to the undo history stack. It's primarily used for synchronized
```

**Review by glendaviesnz — APPROVED** [review]
This tested well for me. Without this there was an additional step in the undo history for pattern overrides, but with the PR the stack has the expected number of entries. 

Although as noted this doesn't yet work with all blocks, given that its usage is currently limited to the new pattern block behind an experimental flag I think it is safe to merge and iterate.

**Comment by kevin940726** [comment]
I'll merge this and we can iterate on the API design in the future. It's currently still private anyway. Thanks for the reviews!
